### PR TITLE
Foundation for Xcode 10 Beta 1-2

### PR DESCRIPTION
**Pull request by chamons**

### CODE REVIEW

File: src/foundation.cs [hunk 1]
```diff
@@ -2274,18 +2274,44 @@ interface NSKeyedUnarchiverDelegate {
...
+		[Deprecated (PlatformName.iOS, 12, 0, message: "Use 'NSKeyedArchiver (bool)' instead.")]
+		[Deprecated (PlatformName.MacOSX, 10, 14, message: "Use 'NSKeyedArchiver (bool)' instead.")]
 		[Export ("initForWritingWithMutableData:")]
 		IntPtr Constructor (NSMutableData data);
 	
+		[Watch (4,0), TV (11,0), Mac (10,13), iOS (11,0)]
+		[Static]
+		[Export ("archivedDataWithRootObject:requiringSecureCoding:error:")]
+		[return: NullAllowed]
+		NSData ArchivedDataWithRootObject (NSObject @object, bool requiresSecureCoding, [NullAllowed] out NSError error);
```
[the remark this answers is not in the document]
Comment (dalexsoto) [reply src/foundation.cs]: Maybe put this inside a XAMCORE_4_0 so we do not forget?

**Comment by stephen-hawley** [comment]
:eyes:

File: src/foundation.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chamons) [reply src/foundation.cs]: I'm not opposed, but i'm just not sure it's worth making making move from ArchivedDataWithRootObject -> GetArchivedData. I'm happy to do so however.

File: src/foundation.cs [hunk 2]
```diff
@@ -10619,11 +10707,32 @@ interface NSValueTransformer {
...
+		NSString SecureUnarchiveFromDataTransformerName { get; }
+	}
+
+	[Watch (5,0), TV (12,0), Mac (10,14, onlyOn64: true), iOS (12,0)]
+	[BaseType (typeof(NSValueTransformer))]
+	interface NSSecureUnarchiveFromDataTransformer
+	{
+		[Static]
+		[Export ("allowedTopLevelClasses", ArgumentSemantic.Copy)]
+		Class[] AllowedTopLevelClasses { get; }
```
Comment (dalexsoto) [inline src/foundation.cs]: Maybe a Wrap to have an array of `Type`

File: src/foundation.cs [hunk 3]
```diff
@@ -5131,6 +5201,24 @@ partial interface NSUserActivity {
...
+		[Export ("eligibleForPrediction")]
+		bool EligibleForPrediction { [Bind ("isEligibleForPrediction")] get; set; }
+
+		[Watch (5, 0), NoTV, NoMac, iOS (12, 0)]
+		[NullAllowed, Export ("persistentIdentifier")]
+		string PersistentIdentifier { get; set; }
+
+		[Watch (5,0), NoTV, NoMac, iOS (12,0)]
+		[Static]
+		[Export ("deleteSavedUserActivitiesWithPersistentIdentifiers:completionHandler:")]
```
Comment (dalexsoto) [inline src/foundation.cs]: Maybe we can add `[Async]` here

File: src/foundation.cs [hunk 4]
```diff
@@ -5131,6 +5201,24 @@ partial interface NSUserActivity {
...
+		void DeleteSavedUserActivities (string[] persistentIdentifiers, Action handler);
+
+		[Watch (5,0), NoTV, NoMac, iOS (12,0)]
+		[Static]
+		[Export ("deleteAllSavedUserActivitiesWithCompletionHandler:")]
```
Comment (dalexsoto) [inline src/foundation.cs]: Same, `[Async]`

**Comment by chamons** [comment]
I added the code review changes and a test for the new [Wrap].

I'm currently only on laptop and can't test it locally.

File: tests/monotouch-test/Foundation/NSKeyedUnarchiverTest.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	{
+		[Test]
+		public void GetUnarchivedObject_TypeWrappers ()
+		{
+			if (!TestRuntime.CheckXcodeVersion (10, 0))
+				Assert.Ignore ("Ignoring GetUnarchivedObject_TypeWrappers as requires new APIs");
+
+			NSDictionary<NSString, NSString> testValues = new NSDictionary<NSString, NSString> ((NSString)"1", (NSString)"a");
+			NSData data = NSKeyedArchiver.ArchivedDataWithRootObject (testValues, true, out NSError error);
+			Assert.IsNull (error);
```
Comment (spouliot) [inline tests/monotouch-test/Foundation/NSKeyedUnarchiverTest.cs]: name the asserts, too many reports (and sometimes bad symbols) makes debugging failures harder than it have to be

File: tests/monotouch-test/Foundation/NSKeyedUnarchiverTest.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+{
+	[TestFixture]
+	[Preserve (AllMembers = true)]
+	public class NSKeyedUnarchiverTest
+	{
+		[Test]
+		public void GetUnarchivedObject_TypeWrappers ()
+		{
+			if (!TestRuntime.CheckXcodeVersion (10, 0))
+				Assert.Ignore ("Ignoring GetUnarchivedObject_TypeWrappers as requires new APIs");
```
Comment (spouliot) [inline tests/monotouch-test/Foundation/NSKeyedUnarchiverTest.cs]: minor: you can assert the Xcode version (shorter than a version check + assert)

File: tests/monotouch-test/Foundation/NSKeyedUnarchiverTest.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+			o = NSKeyedUnarchiver.GetUnarchivedObject (new Type [] { dictionaryType }, data, out error);
+			Assert.IsNotNull (o);
+			Assert.IsNull (error);
+		}
+
+		[Test]
+		public void DataTransformer_AllowedTopLevelTypes_WrapperTests ()
+		{
+			if (!TestRuntime.CheckXcodeVersion (10, 0))
+				Assert.Ignore ("Ignoring DataTransformer_AllowedTopLevelTypes_WrapperTests as requires new APIs");
```
Comment (spouliot) [inline tests/monotouch-test/Foundation/NSKeyedUnarchiverTest.cs]: minor: you can assert the Xcode version (shorter than a version check + assert)

File: src/foundation.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (VincentDondain) [reply src/foundation.cs]: I think we should rename / obsolete bad APIs whenever we can. It happened to me several times already this season. Sometimes you also realize the name was a poor choice thanks to a new API. As a rule of thumb I really believe we should fix old mistakes whenever we can.

File: src/foundation.cs [hunk 8]
```diff
@@ -10619,11 +10709,36 @@ interface NSValueTransformer {
...
+
+		[Watch (5, 0), TV (12, 0), Mac (10, 14, onlyOn64: true), iOS (12, 0)]
+		[Field ("NSSecureUnarchiveFromDataTransformerName")]
+		NSString SecureUnarchiveFromDataTransformerName { get; }
+	}
+
+	[Watch (5,0), TV (12,0), Mac (10,14, onlyOn64: true), iOS (12,0)]
+	[BaseType (typeof(NSValueTransformer))]
+	interface NSSecureUnarchiveFromDataTransformer
+	{
```
Comment (VincentDondain) [inline src/foundation.cs]: Minor: styling, this goes 1 line higher (:

**Comment by chamons** [comment]
Looks like a legit ios failure - DataTransformer_AllowedTopLevelTypes_WrapperTests

**Comment by chamons** [comment]
I'm punting on those bindings until I fix - https://github.com/xamarin/xamarin-macios/issues/4441

This PR has waited long enough already.

**Pull request merged**